language-plutus-core/src/Language/PlutusCore/Constant/Make.hs

```diff
@@ -43,20 +52,30 @@ makeBuiltinBS size bs = checkBoundsBS size bs ? BuiltinBS () size bs
 makeBuiltinSize :: Size -> Size -> Maybe (Constant ())
 makeBuiltinSize size size' = checkBoundsSize size size' ? BuiltinSize () size
 
--- | Coerce a 'Bool' to the corresponding PLC's @boolean@.
-makeDupBuiltinBool :: Bool -> Value TyName Name ()
-makeDupBuiltinBool b = runQuote $ if b then getBuiltinTrue else getBuiltinFalse
+-- | Convert a 'Bool' to the corresponding PLC's @boolean@.
+makeBuiltinBool :: Bool -> Quote (Value TyName Name ())
+makeBuiltinBool b = if b then getBuiltinTrue else getBuiltinFalse
 
--- | Coerce a Haskell value to the corresponding PLC constant indexed by size
+-- | Convert a Haskell value to the corresponding PLC constant indexed by size
 -- checking all constraints (e.g. an 'Integer' is in appropriate bounds) along the way.
 makeSizedConstant :: Size -> TypedBuiltinSized a -> a -> Maybe (Constant ())
 makeSizedConstant size TypedBuiltinSizedInt  int   = makeBuiltinInt  size int
 makeSizedConstant size TypedBuiltinSizedBS   bs    = makeBuiltinBS   size bs
 makeSizedConstant size TypedBuiltinSizedSize size' = makeBuiltinSize size size'
 
--- | Coerce a Haskell value to the corresponding PLC value checking all constraints
+-- | Convert a Haskell value to the corresponding PLC value checking all constraints
 -- (e.g. an 'Integer' is in appropriate bounds) along the way.
-makeConstant :: TypedBuiltin Size a -> a -> Maybe (Value TyName Name ())
-makeConstant (TypedBuiltinSized se tbs) x =
-    Constant () <$> makeSizedConstant (flattenSizeEntry se) tbs x
-makeConstant TypedBuiltinBool           b = Just $ makeDupBuiltinBool b
+-- Does not preserve the global uniqueness condition.
+makeBuiltin :: TypedBuiltinValue Size a -> Quote (Maybe (Value TyName Name ()))
+makeBuiltin (TypedBuiltinValue tb x) = case tb of
+    (TypedBuiltinSized se tbs) ->
+        return $ Constant () <$> makeSizedConstant (flattenSizeEntry se) tbs x
+    TypedBuiltinBool           -> Just <$> makeBuiltinBool x
+
+-- | Convert a Haskell value to a PLC value checking all constraints
+-- (e.g. an 'Integer' is in appropriate bounds) along the way and
+-- fail in case constraints are not satisfied.
+-- Does not preserve the global uniqueness condition.
+unsafeMakeBuiltin :: TypedBuiltinValue Size a -> Value TyName Name ()
```

Comment:
Why not run this in the context of a monad with error handling? I notice that you're using this under `GenT m` later, where you could require a `MonadError` constraint on `m`.

Reply:
Indeed, coudn't you turn things like this into `Alternative.empty` in `GenT`?

Also, since a lot of these things are running in `GenT Quote` later, might it not be possible to just keep the `Quote` constraint?

Reply:
> Indeed, coudn't you turn things like this into `Alternative.empty` in `GenT`?

No, because it just silently discards a generator while I want to throw a true error that is not supposed to be catchable. This is a perfect case to use `error`, so I do not want to complicate.

Reply:
I guess I'm confused - why would it be wrong to just discard the generator in this case? It's produced something invalid, so you should throw the result away. I'm not sure that crashing is better. Or is this supposed to be impossible? If that's so, then I don't think it needs the `unsafe` prefix.

Reply:
> Or is this supposed to be impossible?

Yes.

> If that's so, then I don't think it needs the `unsafe` prefix.

It's an exported function from a module, hence I believe it should have the `unsafe` prefix, because someone else might decide to use it and better be clear here.